pkg/dfns-api-client/auth_transport.go

```diff
@@ -0,0 +1,286 @@
+
+	authHeader, err := strconv.ParseBool(authHeaderStr)
+	if err != nil {
+		return false, fmt.Errorf("error parsing header value as boolean: %w", err)
+	}
+
+	return authHeader, nil
+}
+
+// PerformSimpleRequest performs a simple HTTP request.
+func (auth *authTransport) performSimpleRequest(req *http.Request) (*http.Response, error) {
+	auth.setHeaders(req)
+
+	return executeRequest(req)
+}
+
+// PerformUserActionRequest performs a user action HTTP request.
+func (auth *authTransport) performUserActionRequest(req *http.Request) error {
+	auth.setHeaders(req)
+
+	return auth.executeUserActionRequest(req)
+}
+
+// setHeaders sets the request headers.
+func (auth *authTransport) setHeaders(req *http.Request) {
+	req.Header.Set(appIDHeader, auth.AppId)
+	req.Header.Set(nonceHeader, generateNonce())
+
+	if auth.AppSecret != nil {
+		req.Header.Set(appSecretHeader, *auth.AppSecret)
+	}
+
+	if auth.AuthToken != nil {
+		req.Header.Set(authorizationHeader, "Bearer "+*auth.AuthToken)
+	}
+
+	req.Header.Set(contentTypeHeader, "application/json")
+}
+
+func executeRequest(req *http.Request) (*http.Response, error) {
+	response, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fill Request otherwise it's nil
+	response.Request = req
+
+	if err := handleResponseError(response); err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// handleResponseError handles errors in the HTTP response.
+func handleResponseError(response *http.Response) error {
+	if response.StatusCode < 400 || response.StatusCode == PolicyPendingErrorCode {
+		return nil // No error, return nil
+	}
+
+	respBody, err := io.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %v", err)
+	}
+	defer response.Body.Close()
+
+	var body map[string]interface{}
+	if err := json.Unmarshal(respBody, &body); err != nil {
+		return fmt.Errorf("failed to parse response body: %v", err)
+	}
+
+	if response.StatusCode == PolicyPendingErrorCode {
+		return NewPolicyPendingError(body)
+	}
+
+	var message string
+	if errorObj, ok := body["error"].(map[string]interface{}); ok {
+		if errMsg, ok := errorObj["message"].(string); ok {
+			message = errMsg
+		}
+	} else if errMsg, ok := body["message"].(string); ok {
+		message = errMsg
+	} else {
+		message = "Unknown error"
+	}
+
+	return NewDfnsError(response.StatusCode, message, map[string]interface{}{
+		"URL":     response.Request.URL.String(),
+		"Headers": response.Request.Header,
+		"Body":    body,
+	})
+}
+
+// executeUserActionRequest executes a user action HTTP request.
+func (auth *authTransport) executeUserActionRequest(req *http.Request) error {
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		return err
+	}
+	defer req.Body.Close()
+
+	createUserActionResp, err := auth.createUserActionChallenge(string(body), req.Method, req.URL.Path)
+	if err != nil {
+		return err
+	}
+
+	req.Body = io.NopCloser(bytes.NewReader(body))
```

Comment:
what's the purpose of this line? because we read the body once already?

Reply:
Yes exactly, that's because we already consumed the body when creating the signing challenge (UserActionPayload = body)

If we don't reset the req.Body, it will be nil (And we will have a problem later when doing `http.DefaultClient.Do(req)` later

The `io.NopCloser` is only to satsify the io.ReadCloser interface

Reply:
can this line be moved to either right after line 150 (unless `body` is not accessible after it's wrapped in a reader) or to the end of the function right before return?

it's a bit confusing to be in between  `auth.createUserActionChallenge` and `auth.Signer.Sign` calls when it's not related to either call.